pkg/backend/filestate/backend.go

```diff
@@ -190,22 +243,62 @@ func New(ctx context.Context, d diag.Sink, originalURL string, project *workspac
 		bucket:      wbucket,
 		lockID:      lockID.String(),
 		gzip:        gzipCompression,
-		store:       newLegacyReferenceStore(wbucket),
 	}
 	backend.currentProject.Store(project)
 
-	// Read the Pulumi state metadata
-	// and ensure that it is compatible with this version of the CLI.
+	// Read the Pulumi state metadata, if any.
+	// This will tell us which store to use.
 	meta, err := ensurePulumiMeta(ctx, wbucket)
 	if err != nil {
 		return nil, err
 	}
-	if meta.Version != 0 {
+
+	// projectMode tracks whether the current state supports project-scoped stacks.
+	// Historically, the filestate backend did not support this.
+	// To avoid breaking old stacks, we use legacy mode for existing states.
+	// We use project mode only if one of the following is true:
+	//
+	//  - The state has a .pulumi/Pulumi.yaml file (the 'meta' file)
+	//    and the version is 1 or greater.
+	//  - The state is entirely new so there's no risk of breaking old stacks.
+	//
+	// All actual logic of project mode vs legacy mode is handled by the referenceStore.
+	// This boolean just helps us warn users about unmigrated stacks.
+	var projectMode bool
+	switch meta.Version {
+	case 0:
+		backend.store = newLegacyReferenceStore(wbucket)
+	case 1:
+		backend.store = newProjectReferenceStore(wbucket, backend.currentProject.Load)
+		projectMode = true
+	default:
 		return nil, fmt.Errorf(
 			"state store unsupported: 'Pulumi.yaml' version (%d) is not supported "+
 				"by this version of the Pulumi CLI", meta.Version)
 	}
 
+	// If we're not in project mode, or we've disabled the warning, we're done.
+	if !projectMode || cmdutil.IsTruthy(os.Getenv(PulumiFilestateNoLegacyWarningEnvVar)) {
```

Comment:
If you add it to `sdk/go/common/env/env.go` you can do
```go
	if !projectMode || env.PulumiFilestateNoLegacyWarningEnvVar.Value()) {
```

Reply:
If we do this, we'll have to drop #12497 and switch to t.Setenv for tests again.
I have some vague ideas on how we could make util/env-based code more testable, but I would suggest we defer that for now. WDYT?
(If you feel strongly about this, I'm open to abandoning #12497.)

Reply:
Ah. Maybe we can just add it to the `env.go` file and then add a TODO there.